File: apps/watchos/VaiMarciaWatch Watch App/Playback/PlaybackController.swift
import Combine
import Foundation

/// Único ponto de entrada que a camada de apresentação chama ao tocar em um botão. Guarda a
/// estratégia já resolvida (definida na inicialização / mudança de pareamento, veja
/// `PlaybackStrategyResolver`) e roteia para o engine correspondente — sem cálculo de
/// estratégia a cada toque.
final class PlaybackController: ObservableObject {
    static let shared = PlaybackController()

    @Published private(set) var strategy: PlaybackStrategy = .phoneOnly

    private let resolver: PlaybackStrategyResolver
    private let directEngine: PlaybackEngine
    private let relayDispatcher: PlaybackEngine

    init(
        resolver: PlaybackStrategyResolver = PlaybackStrategyResolver(),
        directEngine: PlaybackEngine = DirectPlaybackEngine(),
        relayDispatcher: PlaybackEngine = RelayPlaybackDispatcher(),
    ) {
        self.resolver = resolver
        self.directEngine = directEngine
        self.relayDispatcher = relayDispatcher
        refreshStrategy()
    }

    func refreshStrategy() {
        strategy = resolver.resolve()
    }

    func play(audioId: String, completion: @escaping (PlaybackResult) -> Void = { _ in }) {
        switch strategy {
        case .direct:
            directEngine.play(audioId: audioId, completion: completion)
        case .relay:
            relayDispatcher.play(audioId: audioId, completion: completion)
        case .phoneOnly:
            completion(.failed(reason: "phone_only_no_watch_output"))
        }
    }
}

File: apps/wearos/app/src/main/java/com/togplay/vaimarcia/wear/playback/PlaybackController.kt
package com.togplay.vaimarcia.wear.playback

import com.togplay.vaimarcia.wear.domain.PlaybackStrategy
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Único ponto de entrada que a camada de apresentação chama ao tocar em um botão. Guarda a
 * estratégia já resolvida (definida na inicialização / mudança de pareamento, veja
 * PlaybackStrategyResolver) e roteia para o engine correspondente — sem cálculo de estratégia
 * a cada toque.
 */
@Singleton
class PlaybackController @Inject constructor(
    private val strategyResolver: PlaybackStrategyResolver,
    private val directEngine: DirectPlaybackEngine,
    private val relayDispatcher: RelayPlaybackDispatcher,
) {
    private val _strategy = MutableStateFlow(PlaybackStrategy.PHONE_ONLY)
    val strategy: StateFlow<PlaybackStrategy> = _strategy.asStateFlow()

    suspend fun refreshStrategy() {
        _strategy.value = strategyResolver.resolve()
    }

    suspend fun play(audioId: String): PlaybackResult = when (_strategy.value) {
        PlaybackStrategy.DIRECT -> directEngine.play(audioId)
        PlaybackStrategy.RELAY -> relayDispatcher.play(audioId)
        PlaybackStrategy.PHONE_ONLY -> PlaybackResult.Failed("phone_only_no_watch_output")
    }
}
